s3/src/main/scala/com/meltwater/fawn/s3/S3.scala

```diff
@@ -0,0 +1,439 @@
+  def listMultipartUploads(
+      bucket: String,
+      optHeaders: Headers = Headers.empty): F[ListMultipartUploadsResponse]
+  def listParts(
+      bucket: String,
+      key: String,
+      uploadId: String,
+      optHeaders: Headers = Headers.empty): F[ListPartsResponse]
+  def uploadPart[T](
+      bucket: String,
+      key: String,
+      partNumber: Int,
+      uploadId: String,
+      t: T,
+      optHeaders: Headers = Headers.empty)(implicit enc: EntityEncoder[F, T]): F[UploadPartResponse]
+  def completeMultipartUpload(
+      bucket: String,
+      key: String,
+      uploadId: String,
+      parts: List[String],
+      optHeaders: Headers = Headers.empty): F[CompleteMultipartUploadResponse]
+  def startMultipartUpload(bucket: String, key: String): Resource[F, MultipartUpload[F]]
+}
+
+object S3 {
+
+  implicit def xtractDecoder[F[_]: Sync, T: XmlReader]: EntityDecoder[F, T] = xml[F].transform {
+    r =>
+      r.flatMap { elem =>
+        XmlReader.of[T].read(elem) match {
+          case ParseSuccess(get)              => get.asRight
+          case PartialParseSuccess(_, errors) =>
+            MalformedMessageBodyFailure(s"Unable to parse body: $errors").asLeft
+          case ParseFailure(errors)           =>
+            MalformedMessageBodyFailure(s"Unable to parse body: $errors").asLeft
+        }
+      }
+  }
+
+  case class S3Error(code: Status, headers: Headers, body: String)
+      extends Throwable
+      with NoStackTrace {
+    override def getMessage: String =
+      s"S3Error(code = $code, headers = $headers, body = \"$body\")"
+  }
+
+  case class HeaderError(name: String, headers: Headers)
+      extends Throwable(s"Could not find header $name")
+      with NoStackTrace {
+    override def toString = s"HeaderError(name = $name)"
+  }
+
+  def apply[F[_]: Sync: Clock: Logger](
+      baseClient: Client[F],
+      credentials: AWSCredentials,
+      awsRegion: AWSRegion): S3[F] = new S3[F] {
+
+    private val client: Client[F] =
+      V4Middleware[F](credentials, awsRegion, AWSService.s3).apply(baseClient)
+
+    private val region: String = awsRegion.value
+
+    private def handleError(r: Response[F]): F[Throwable] =
+      r.bodyText.compile.foldMonoid.map { b =>
+        S3Error(r.status, r.headers, b)
+      }
+
+    private def getHeader(key: String, headers: Headers): F[Header] =
+      Sync[F].fromEither(headers.get(key.ci).toRight(HeaderError(key, headers)))
+
+    private def createBucketBody: Elem =
+      <CreateBucketConfiguration xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
+        <LocationConstraint>{region}</LocationConstraint>
+      </CreateBucketConfiguration>
+
+    override def createBucket(
+        bucket: String,
+        optHeaders: Headers = Headers.empty): F[CreateBucketResponse] =
+      client
+        .run(
+          Request[F](
+            Method.PUT,
+            Uri.unsafeFromString(s"https://$bucket.s3.$region.amazonaws.com"),
```

Comment:
the `amazonaws.com` bit should be part of an `endpoint` configuration option that would take `https://s3.amazonaws.com` or `https://nyc3.digitaloceanspaces.com` for example

Reply:
I'll have to remove the use of `AWSRegion` and change it to a `String` since the region codes are different

Reply:
the region codes are actually the same for digitalocean are actually the same

Reply:
I've added `host` as an input in the apply method for `S3` which should allow us to choose. The only thing is the region will need to be added to the `host` in the correct location since these differ between hosts.

Reply:
we should probably try and mirror the official clients and use an `endpoint` argument

Reply:
https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/AmazonS3.java#L249-L269
it looks like they just parse the subdomain and the host out and then use those, which would be fine